src/rule_dsl.jl

```diff
@@ -212,22 +213,26 @@ a RuleSet until there are no changes.
 """
 struct RuleSet <: AbstractRule
     rules::Vector{AbstractRule}
+    applyall::Bool
+    recurse::Bool
 end
+RuleSet(rules; applyall=false, recurse=true) = RuleSet(rules, applyall, recurse)
+
 
 struct RuleRewriteError
     rule
     expr
 end
 
-function (r::RuleSet)(term; depth=typemax(Int))
+function (r::RuleSet)(@nospecialize(term); depth=typemax(Int))
     rules = r.rules
     term = to_symbolic(term)
     # simplify the subexpressions
     if depth == 0
         return term
     end
     if term isa Symbolic
-        if term isa Term
+        if term isa Term && r.recurse
```

Comment:
why would you not want to recurse?

Reply:
Oh I see what's up. So you want to recurse only from the topmost RuleSet

Reply:
But it does not seem like it should be a flag on the RuleSet type but should be an internal argument to this function. It should start off with true and become false if going into a nested RuleSet.

Reply:
That would leave the `RuleSet` objects free to be composed however without worrying which one should go on top. Tell me if I'm talking nonsense!

Reply:
Yeah, you're tight, I'll change this.